### PR TITLE
[CPU] Optimize small batch case for PagedAttention

### PR DESCRIPTION
**Pull request by luo-cheng2021**
### Details:
 - *Generate more work items to avoid thread imbalance*
 - *...*

### Tickets:
 - *[156347](https://jira.devtools.intel.com/browse/CVS-156347)*
 - *[158477](https://jira.devtools.intel.com/browse/CVS-158477)*

### CODE REVIEW

**Comment by dmitry-gorokhov** [comment]
@zhangYiIntel Could you please take a look?

File: src/plugins/intel_cpu/src/nodes/kernels/scaled_attn/executor_pa.cpp [hunk 1]
```diff
@@ -1162,33 +1163,58 @@ struct MHAHelper {
...
+                if (past_lens.ptr<int32_t>()[b] != kv_len)
+                    prefer_static_loop = false;
+            }
+        } else {
+            // bigger batch, probably it's vllm path, skip the test to save the cost
+            prefer_static_loop = false;
+        }
+        auto loop_qk = [&](size_t b, size_t pk_in_blocks, size_t hx) {
             auto context_len = static_cast<size_t>(past_lens.ptr<int32_t>()[b]) + 1;
+            size_t hk, hq_beg, hq_end;
```
Comment (zhangYiIntel) [inline src/plugins/intel_cpu/src/nodes/kernels/scaled_attn/executor_pa.cpp:1183]: The setting of `hq_beg` and `hg_end` are the same between `loop_qk` and 'loop_wk', please consider combine them together to a common place

File: src/plugins/intel_cpu/src/nodes/kernels/scaled_attn/executor_pa.cpp [hunk 2]
```diff
@@ -1210,7 +1236,14 @@ struct MHAHelper {
                                        ov::element::f32,
                                        ov::element::f32,
                                        alibi_slope);
-        });
+        };
+        if (prefer_static_loop) {
+            parallel_for3d(B, kv_len_in_blocks, loop_hk ? _Hk : _H, loop_qk);
```
Comment (zhangYiIntel) [inline src/plugins/intel_cpu/src/nodes/kernels/scaled_attn/executor_pa.cpp:1241]: The last dim of `parallel_for` is the same for several `parallel_for`, please consider to set them in a common place.
```
size_t h_dims = loop_hk ? _Hk : _H, loop_qk
```

File: src/plugins/intel_cpu/src/nodes/kernels/scaled_attn/executor_pa.cpp [hunk 3]
```diff
@@ -1416,7 +1465,23 @@ struct MHA {
             }
         });
 
-        parallel_for2d_dynamic(attn_work_count, Hk, [&](size_t w, size_t hk) {
+        // loop along HK dimension: if mixed first/second token and elements count is enough, loop HK to reuse KV in the CPU cache
+        //    else if elements count is small, prefer to loop H to get more work to avoid thread imbalance
+        bool loop_hk = _workitems.get_reorder_max_batch_size() == past_lens.m_dims[0] ||        // if only first token, loop H
+            attn_work_count * Hk <= 2 * _helper._nthr ? false : true;                           // or less than 2 work items per thread, loop H
```
Comment (zhangYiIntel) [inline src/plugins/intel_cpu/src/nodes/kernels/scaled_attn/executor_pa.cpp:1471]: Please align the threshold between `attn_work_count * Hk <= 2 * _helper._nthr` in `exec_loop_mixed` and `bool loop_hk = B * kv_len_in_blocks * _Hk > 2 * _nthr` in `exec_loop_bhl`

Comment (zhangYiIntel) [reply src/plugins/intel_cpu/src/nodes/kernels/scaled_attn/executor_pa.cpp:1471]: Please also apply the threshold comment to corresponding place in `exec_loop_bhl`

**Review by zhangYiIntel — APPROVED** [review]
General LGTM. Minor changes to reduce redundant code.

File: src/plugins/intel_cpu/src/nodes/kernels/scaled_attn/executor_pa.cpp (hunk 1 above)
Comment (luo-cheng2021) [reply src/plugins/intel_cpu/src/nodes/kernels/scaled_attn/executor_pa.cpp]: Done.

File: src/plugins/intel_cpu/src/nodes/kernels/scaled_attn/executor_pa.cpp (hunk 2 above)
Comment (luo-cheng2021) [reply src/plugins/intel_cpu/src/nodes/kernels/scaled_attn/executor_pa.cpp]: Done.

File: src/plugins/intel_cpu/src/nodes/kernels/scaled_attn/executor_pa.cpp (hunk 3 above)
Comment (luo-cheng2021) [reply src/plugins/intel_cpu/src/nodes/kernels/scaled_attn/executor_pa.cpp:1471]: Done.

File: src/plugins/intel_cpu/src/nodes/kernels/scaled_attn/executor_pa.cpp [hunk 4]
```diff
@@ -1162,33 +1163,63 @@ struct MHAHelper {
...
+        if (B <= 32) {
+            prefer_static_loop = true;
+            // small batch and all batch size is same(like SDPA case)
+            auto kv_len = past_lens.ptr<int32_t>()[0];
+            for (size_t b = 1; b < B; b++) {
+                if (past_lens.ptr<int32_t>()[b] != kv_len)
+                    prefer_static_loop = false;
+            }
+        } else {
+            // bigger batch, probably it's vllm path, skip the test to save the cost
```
Comment (dmitry-gorokhov) [inline src/plugins/intel_cpu/src/nodes/kernels/scaled_attn/executor_pa.cpp:1179]: mentioning VLLM here doesn't makes amy sense here, We might have different sources of big batch.

Comment (luo-cheng2021) [reply src/plugins/intel_cpu/src/nodes/kernels/scaled_attn/executor_pa.cpp]: Done.

**Review by dmitry-gorokhov — APPROVED** [review]

**Pull request merged**